src/phongo_bson_encode.c

```diff
@@ -451,26 +474,47 @@ static void php_phongo_zval_to_bson_internal(zval* data, php_phongo_field_path*
 					return;
 				}
 
+				if (instanceof_function(Z_OBJCE_P(data), php_phongo_persistable_ce)) {
+					if (Z_TYPE(obj_data) == IS_OBJECT && instanceof_function(Z_OBJCE(obj_data), php_phongo_packedarray_ce)) {
+						phongo_throw_exception(
+							PHONGO_ERROR_UNEXPECTED_VALUE,
+							"Expected %s::%s() to return an array, stdClass, or %s, %s given",
+							ZSTR_VAL(Z_OBJCE_P(data)->name),
+							BSON_SERIALIZE_FUNC_NAME,
+							ZSTR_VAL(php_phongo_document_ce->name),
+							PHONGO_ZVAL_CLASS_OR_TYPE_NAME(obj_data));
+
+						goto cleanup;
+					}
+
+					bson_append_binary(bson, PHONGO_ODM_FIELD_NAME, -1, 0x80, (const uint8_t*) Z_OBJCE_P(data)->name->val, Z_OBJCE_P(data)->name->len);
+					/* Ensure that we ignore an existing key with the same name
+					 * if one exists in the bsonSerialize() return value. */
+					skip_odm_field = true;
+				}
+
+				// If bsonSerialize() returns a BSON document or packedArray instance, recurse to copy data over directly
+				if (Z_TYPE(obj_data) == IS_OBJECT && (instanceof_function(Z_OBJCE(obj_data), php_phongo_document_ce) || instanceof_function(Z_OBJCE(obj_data), php_phongo_packedarray_ce))) {
+					php_phongo_zval_to_bson_internal(&obj_data, field_path, flags, bson, bson_out);
+
+					goto done;
+				}
+
 				if (Z_TYPE(obj_data) != IS_ARRAY && !(Z_TYPE(obj_data) == IS_OBJECT && instanceof_function(Z_OBJCE(obj_data), zend_standard_class_def))) {
 					phongo_throw_exception(
 						PHONGO_ERROR_UNEXPECTED_VALUE,
-						"Expected %s::%s() to return an array or stdClass, %s given",
+						"Expected %s::%s() to return an array, stdClass, %s, or %s, %s given",
 						ZSTR_VAL(Z_OBJCE_P(data)->name),
 						BSON_SERIALIZE_FUNC_NAME,
+						ZSTR_VAL(php_phongo_document_ce->name),
+						ZSTR_VAL(php_phongo_packedarray_ce->name),
```

Comment:
This logic all looks correct, although it's a bit more convoluted than the type checking we have in `php_phongo_bson_append_object`. This function might also benefit from a comment like:

```
// TODO PHP_VERSION_ID < 80000: obsolete once the tentative return type of bsonSerialize() is enforced
```

For now, I'd just suggest adding a comment. I think we can leave the code as-is since it hopefully won't stick around much longer (or at least we're unlikely to revisit these functions anytime soon).

If I was implementing this from scratch, I'd probably create a utility function just to check the return type of `bsonSerialize()` and pass in the original object's `zend_class_entry*` to control the logic enforced (really just between persistable or serializable). Something like:

```c
/* Checks the return type of a bsonSerialize() method. Returns true on
 * success; otherwise, throws an exception and returns false.
 *
 * TODO: obsolete once PHP 8.0+ is required and tentative return type
 * is enforced. */
static inline bool phongo_check_bson_serialize_return_type(zval *retval, zend_class_entry* ce)
{
    if (ce instanceof persistable) {
        // check types, throw and return false if needed

        return true;
    }

    if (ce instanceof serializable) {       
        // check types, throw and return false if needed

        return true;
    }

    /* consider throwing PHONGO_ERROR_UNEXPECTED_VALUE since we shouldn't
     * get here, but if you want to just return true here you can remove
     * the true returns above and use if/else instead.
     */ 
}
```

Reply:
Added the helper as suggested, that does indeed clean this up a bit.